tracer/src/Datadog.Trace/IAST/IastModule.cs

```diff
@@ -0,0 +1,82 @@
+// </copyright>
+
+#nullable enable
+
+using System;
+using System.Diagnostics;
+using System.Linq;
+using Datadog.Trace.Configuration;
+using Datadog.Trace.Iast;
+using Datadog.Trace.Util;
+
+namespace Datadog.Trace.Iast
+{
+    internal class IastModule
+    {
+        private const string OperationNameHash = "weak_hashing";
+        private const string ServiceHash = "weak_hashing";
+
+        public IastModule()
+        {
+        }
+
+        public static Scope? OnHashingAlgorithm(string? algorithm, IntegrationId integrationId, Datadog.Trace.Iast.Iast iast)
+        {
+            if (algorithm == null || !InvalidHashAlgorithm(algorithm, iast))
+            {
+                return null;
+            }
+
+            return GetScope(algorithm, integrationId);
+        }
+
+        private static Scope GetScope(string evidenceValue, IntegrationId integrationId)
+        {
+            var tracer = Tracer.Instance;
+            var frame = StackWalker.GetFrame();
+            // Sometimes we do not have the file/line but we have the method/class.
+            var filename = frame?.GetFileName();
+            var vulnerability = new Vulnerability(VulnerabilityType.WEAK_HASH, new Location(filename ?? GetMethodName(frame), filename != null ? frame?.GetFileLineNumber() : null), new Evidence(evidenceValue));
+            // The VulnerabilityBatch class is not very useful right now, but we will need it when handling requests
+            var batch = new VulnerabilityBatch();
+            batch.Add(vulnerability);
+
+            // Right now, we always set the IastEnabled tag to "1", but in the future, it might be zero to indicate that a request has not been analyzed
+            var tags = new IastTags()
+            {
+                IastJson = batch.ToString(),
+                IastEnabled = "1"
+            };
+
+            var serviceName = tracer.Settings.GetServiceName(tracer, ServiceHash);
+            var scope = tracer.StartActiveInternal(OperationNameHash, serviceName: serviceName, tags: tags);
+            tracer.TracerManager.Telemetry.IntegrationGeneratedSpan(integrationId);
+
+            return scope;
+        }
+
+        private static string? GetMethodName(StackFrame? frame)
+        {
+            var method = frame?.GetMethod();
+            var declaringType = method?.DeclaringType;
+            var namespaceName = declaringType?.Namespace;
+            var typeName = declaringType?.Name;
+            var methodName = method?.Name;
+
+            if (methodName == null || typeName == null || namespaceName == null)
+            {
+                return null;
+            }
+
+            return $"{namespaceName}.{typeName}::{methodName}";
+        }
+
+        private static bool InvalidHashAlgorithm(string algorithm, Iast iast)
+        {
+            return iast.Settings.InsecureHashingAlgorithmsArray.Contains(algorithm);
```

Comment:
TL;DR `InsecureHashingAlgorithmsArray` is an array, so this is using LINQ because there is no `Array.Contains()`. That is, it's calling the `Enumerable.Contains()` extension method. We generally avoid using LINQ for performance reasons. Also, you probably want a case-insensitive comparison, and `Contains()` will be case-sensitive.

----

At first, I thought this would allocate an enumerator because the array is cast to `IEnumerable<string>` for LINQ, so I looked at the .NET source code. Turns out that LINQ checks if the `IEnumerable<T>` also implements `ICollection<T>` and if so, it returns `ICollection<T>.Contains()`. Array implements `ICollection<T>.Contains()`  and internally it uses `Array.IndexOf()`, so no enumerator is allocated, which is good.

However, this still incurs several methods calls and bounds checking, so I could avoid using LINQ. I suggest a simple `foreach` loop instead:
```suggestion
            foreach (var insecureAlgorithm in iast.Settings.InsecureHashingAlgorithmsArray)
            {
                if (string.Equals(algorithm, insecureAlgorithm, StringComparison.OrdinalIgnoreCase))
                {
                    return true;
                }
            }
            
            return false;
```

Reply:
Cool explanation. I will use your code.